README.md
dl-number-validation - A python library for validation of USA drivers license numbers
=========================
Validates if a drivers license is a valid format for a given two letter state abbreviation according to http://adr-inc.com/PDFs/State_DLFormats.pdf

Installation
------------
`pip install dl-number-validation`

Usage
-----
```
from dl-number-validation import is_valid

if is_valid('C1234567', 'CA'):
    print "Your drivers license number is valid! Yay!"
```

If you want US territories to be used for validation call `is_valid` with `allow_territories=True`. Be warned, currently these additional territories accept any format of drivers license.


CHANGES.txt
v0.1.0, 2015/02/18 -- Initial release.
v0.1.4, 2015/02/25 -- Fixed Virginia's format.
v0.2.0, 2015/03/09 -- New Source for formats.
v0.2.1, 2015/08/25 -- Added format for Oregon
v0.3.0, 2015/09/02 -- Included more possible formats.
v0.3.1, 2015/12/03 -- Fixed West Virginia's format.
v0.3.2, 2019/08/12 -- Add support for new format in Washington state.

test_dlnvalidation.py
import unittest
from dlnvalidation import (
    STATE_FORMATS, OTHER_STATES, is_valid)


class DLNValidationTest(unittest.TestCase):

    def test_invalid_dl_number_all(self):
        """
        Invalid formats for every state.
        """
        for dl_state in STATE_FORMATS.keys():
            self.assertFalse(is_valid('!@#$%^&*()', dl_state))
            self.assertFalse(is_valid(
                'reallyreallyreallylooooooonyyyalphastring',
                dl_state))
            self.assertFalse(is_valid(
                '00000001111111112222222222222222222222222',
                dl_state))
            self.assertFalse(is_valid('a012345678901234', dl_state))
            self.assertFalse(is_valid('', dl_state))
            self.assertFalse(is_valid('-', dl_state))
            self.assertFalse(is_valid('*', dl_state))
            self.assertFalse(is_valid('+', dl_state))
            self.assertFalse(is_valid('#', dl_state))
            self.assertFalse(is_valid('&', dl_state))
            self.assertFalse(is_valid('~', dl_state))
            self.assertFalse(is_valid('123~4567', dl_state))
            self.assertFalse(is_valid('@', dl_state))
            self.assertFalse(is_valid('abcd', dl_state))
            self.assertFalse(is_valid('1234a67890b23', dl_state))

    def test_alabama(self):
        dl_state = 'AL'
        self.assertFalse(is_valid('1', dl_state))
        self.assertTrue(is_valid('1234567', dl_state))
        self.assertTrue(is_valid('ab34c6', dl_state))
        self.assertFalse(is_valid('a123456', dl_state))

    def test_alaska(self):
        dl_state = 'AK'
        self.assertTrue(is_valid('1', dl_state))
        self.assertTrue(is_valid('1234567', dl_state))
        self.assertFalse(is_valid('a123456', dl_state))

    def test_arizona(self):
        dl_state = 'AZ'
        self.assertFalse(is_valid('1234567', dl_state))
        self.assertTrue(is_valid('a12345678', dl_state))
        self.assertTrue(is_valid('A12345678', dl_state))
        self.assertFalse(is_valid('aB12345', dl_state))
        self.assertTrue(is_valid('123456789', dl_state))

    def test_arkansas(self):
        dl_state = 'AR'
        self.assertTrue(is_valid('12345678', dl_state))
        self.assertTrue(is_valid('123456789', dl_state))
        self.assertFalse(is_valid('a123456', dl_state))

    def test_california(self):
        dl_state = 'CA'
        self.assertFalse(is_valid('1234567', dl_state))
        self.assertTrue(is_valid('a1234567', dl_state))
        self.assertTrue(is_valid('A1234567', dl_state))
        self.assertFalse(is_valid('ab1234567', dl_state))

    def test_colorado(self):
        dl_state = 'CO'
        self.assertTrue(is_valid('123456789', dl_state))
        self.assertFalse(is_valid('12345678', dl_state))
        self.assertTrue(is_valid('a123456', dl_state))
        self.assertTrue(is_valid('A123', dl_state))
        self.assertFalse(is_valid('Ab12345', dl_state))

    def test_connecticut(self):
        dl_state = 'CT'
        self.assertTrue(is_valid('123456789', dl_state))
        self.assertFalse(is_valid('1234567', dl_state))
        self.assertFalse(is_valid('a123456', dl_state))

    def test_delaware(self):
        dl_state = 'DE'
        self.assertTrue(is_valid('1', dl_state))
        self.assertTrue(is_valid('1234567', dl_state))
        self.assertFalse(is_valid('a123456', dl_state))

    def test_dc(self):
        dl_state = 'DC'
        self.assertFalse(is_valid('1234', dl_state))
        self.assertTrue(is_valid('1234567', dl_state))
        self.assertTrue(is_valid('123456789', dl_state))
        self.assertFalse(is_valid('a123456', dl_state))

    def test_florida(self):
        dl_state = 'FL'
        self.assertFalse(is_valid('1234567', dl_state))
        self.assertTrue(is_valid('a123456789012', dl_state))
        self.assertTrue(is_valid('C123456789012', dl_state))
        self.assertTrue(is_valid('a12', dl_state))
        self.assertTrue(is_valid('a1234567890', dl_state))
        self.assertFalse(is_valid('ab1234567', dl_state))

    def test_georgia(self):
        dl_state = 'GA'
        self.assertFalse(is_valid('12345', dl_state))
        self.assertTrue(is_valid('1234567', dl_state))
        self.assertTrue(is_valid('123456789', dl_state))
        self.assertFalse(is_valid('a123456', dl_state))

    def test_hawaii(self):
        dl_state = 'HI'
        self.assertTrue(is_valid('123456789', dl_state))
        self.assertTrue(is_valid('h12345678', dl_state))
        self.assertFalse(is_valid('ab1234567', dl_state))
        self.assertFalse(is_valid('ab1234', dl_state))

    def test_idaho(self):
        dl_state = 'ID'
        self.assertTrue(is_valid('123456789', dl_state))
        self.assertTrue(is_valid('ab123456a', dl_state))
        self.assertTrue(is_valid('AB123456A', dl_state))
        self.assertFalse(is_valid('a12345678', dl_state))
        self.assertFalse(is_valid('ab1234', dl_state))

    def test_illinois(self):
        dl_state = 'IL'
        self.assertFalse(is_valid('123456789', dl_state))
        self.assertTrue(is_valid('a12345678901', dl_state))
        self.assertFalse(is_valid('A123456789012', dl_state))
        self.assertFalse(is_valid('ab1234567', dl_state))

    def test_indiana(self):
        dl_state = 'IN'
        self.assertTrue(is_valid('1234567890', dl_state))
        self.assertTrue(is_valid('a123456789', dl_state))
        self.assertTrue(is_valid('A123456789', dl_state))
        self.assertFalse(is_valid('ab1234567', dl_state))

    def test_iowa(self):
        dl_state = 'IA'
        self.assertTrue(is_valid('123456789', dl_state))
        self.assertFalse(is_valid('a12345678', dl_state))
        self.assertFalse(is_valid('ab1234567', dl_state))
        self.assertTrue(is_valid('123ab1234', 'IA'))
        self.assertTrue(is_valid('123AB1234', 'IA'))

    def test_kansas(self):
        dl_state = 'KS'
        self.assertTrue(is_valid('123456789', dl_state))
        self.assertTrue(is_valid('k12345678', dl_state))
        self.assertTrue(is_valid('K12345678', dl_state))
        self.assertFalse(is_valid('a12345678', dl_state))
        self.assertFalse(is_valid('ab1234567', dl_state))
        self.assertFalse(is_valid('a1a1a', dl_state))

    def test_kentucky(self):
        dl_state = 'KY'
        self.assertTrue(is_valid('k12345678', dl_state))
        self.assertTrue(is_valid('a12345678', dl_state))
        self.assertTrue(is_valid('A12345678', dl_state))
        self.assertFalse(is_valid('ab1234567', dl_state))

    def test_louisiana(self):
        dl_state = 'LA'
        self.assertTrue(is_valid('123456789', dl_state))
        self.assertFalse(is_valid('1234', dl_state))
        self.assertFalse(is_valid('a123456', dl_state))

    def test_maine(self):
        dl_state = 'ME'
        self.assertFalse(is_valid('1234', dl_state))
        self.assertTrue(is_valid('1234567', dl_state))
        self.assertFalse(is_valid('1234567a', dl_state))
        self.assertFalse(is_valid('a123456', dl_state))

    def test_maryland(self):
        dl_state = 'MD'
        self.assertFalse(is_valid('1234567', dl_state))
        self.assertTrue(is_valid('a123456789012', dl_state))
        self.assertTrue(is_valid('M123456789012', dl_state))
        self.assertFalse(is_valid('a1234567890', dl_state))
        self.assertFalse(is_valid('ab1234567', dl_state))

    def test_massachusetts(self):
        dl_state = 'MA'
        self.assertTrue(is_valid('123456789', dl_state))
        self.assertTrue(is_valid('s12345678', dl_state))
        self.assertTrue(is_valid('S12345678', dl_state))
        self.assertFalse(is_valid('a1234567890', dl_state))
        self.assertFalse(is_valid('ab1234567', dl_state))

    def test_michigan(self):
        dl_state = 'MI'
        self.assertFalse(is_valid('1234567', dl_state))
        self.assertTrue(is_valid('a123456789012', dl_state))
        self.assertTrue(is_valid('L123456789012', dl_state))
        self.assertFalse(is_valid('a1234567890', dl_state))
        self.assertFalse(is_valid('ab1234567', dl_state))

    def test_minnesota(self):
        dl_state = 'MN'
        self.assertFalse(is_valid('1234567', dl_state))
        self.assertTrue(is_valid('a123456789012', dl_state))
        self.assertTrue(is_valid('M123456789012', dl_state))
        self.assertFalse(is_valid('a1234567890', dl_state))
        self.assertFalse(is_valid('ab1234567', dl_state))

    def test_mississippi(self):
        dl_state = 'MS'
        self.assertTrue(is_valid('123456789', dl_state))
        self.assertFalse(is_valid('a123456789012', dl_state))
        self.assertFalse(is_valid('a1234567890', dl_state))
        self.assertFalse(is_valid('ab1234567', dl_state))

    def test_missouri(self):
        dl_state = 'MO'
        self.assertTrue(is_valid('123456789', dl_state))
        self.assertTrue(is_valid('a12345678', dl_state))
        self.assertTrue(is_valid('a12345', dl_state))
        self.assertTrue(is_valid('A12345678', dl_state))
        self.assertTrue(is_valid('A123456', dl_state))

        self.assertFalse(is_valid('12345678ab', dl_state))
        self.assertFalse(is_valid('123456789a', dl_state))
        self.assertFalse(is_valid('ab1234567', dl_state))

    def test_montana(self):
        dl_state = 'MT'
        self.assertTrue(is_valid('abcdefghi', dl_state))
        self.assertTrue(is_valid('1234567890123', dl_state))
        self.assertFalse(is_valid('a12345678', dl_state))
        self.assertFalse(is_valid('a1234567890', dl_state))
        self.assertFalse(is_valid('ab1234567', dl_state))

    def test_nebraska(self):
        dl_state = 'NE'
        self.assertTrue(is_valid('a123456', dl_state))
        self.assertTrue(is_valid('a123', dl_state))
        self.assertTrue(is_valid('a12345678', dl_state))
        self.assertFalse(is_valid('12345678', dl_state))
        self.assertFalse(is_valid('12345678a', dl_state))

    def test_nevada(self):
        dl_state = 'NV'
        self.assertFalse(is_valid('1234', dl_state))
        self.assertTrue(is_valid('1234567890', dl_state))
        self.assertTrue(is_valid('x12345678', dl_state))
        self.assertTrue(is_valid('X12345678', dl_state))
        self.assertFalse(is_valid('a123456', dl_state))
        self.assertFalse(is_valid('a12345678', dl_state))
        self.assertFalse(is_valid('12345678', dl_state))

    def test_new_hampshire(self):
        dl_state = 'NH'
        self.assertFalse(is_valid('123456789', dl_state))
        self.assertFalse(is_valid('a12345', dl_state))
        self.assertFalse(is_valid('ab1234567', dl_state))
        self.assertTrue(is_valid('12abc12345', dl_state))
        self.assertTrue(is_valid('12ABC12345', dl_state))

    def test_new_jersey(self):
        dl_state = 'NJ'
        self.assertFalse(is_valid('123456789', dl_state))
        self.assertTrue(is_valid('a12345678901234', dl_state))
        self.assertTrue(is_valid('N12345678901234', dl_state))
        self.assertFalse(is_valid('a1234567890', dl_state))
        self.assertFalse(is_valid('ab1234567', dl_state))

    def test_new_mexico(self):
        dl_state = 'NM'
        self.assertTrue(is_valid('123456789', dl_state))
        self.assertFalse(is_valid('12345678', dl_state))
        self.assertFalse(is_valid('a123456', dl_state))
        self.assertFalse(is_valid('123456', dl_state))

    def test_new_york(self):
        dl_state = 'NY'
        self.assertTrue(is_valid('123456789', dl_state))
        self.assertTrue(is_valid('A123456789012345678', dl_state))
        self.assertFalse(is_valid('12345678', dl_state))
        self.assertFalse(is_valid('1234567890123456', dl_state))
        self.assertFalse(is_valid('a1234567', dl_state))
        self.assertFalse(is_valid('NEWaYORK', dl_state))
        self.assertFalse(is_valid('123456', dl_state))

    def test_north_carolina(self):
        dl_state = 'NC'
        self.assertTrue(is_valid('123', dl_state))
        self.assertTrue(is_valid('123456789012', dl_state))
        self.assertFalse(is_valid('a123456', dl_state))
        self.assertFalse(is_valid('1234567890123', dl_state))

    def test_north_dakota(self):
        dl_state = 'ND'
        self.assertFalse(is_valid('123', dl_state))
        self.assertTrue(is_valid('123456789', dl_state))
        self.assertFalse(is_valid('a123456', dl_state))
        self.assertFalse(is_valid('1234567890123', dl_state))
        self.assertTrue(is_valid('abc123456', dl_state))
        self.assertTrue(is_valid('NDc123456', dl_state))

    def test_ohio(self):
        dl_state = 'OH'
        self.assertTrue(is_valid('AB123456', dl_state))
        self.assertFalse(is_valid('OH123', dl_state))
        self.assertFalse(is_valid('1234567890123', dl_state))

    def test_oklahoma(self):
        dl_state = 'OK'
        self.assertTrue(is_valid('123456789', dl_state))
        self.assertFalse(is_valid('ab123', dl_state))
        self.assertTrue(is_valid('a123456789', dl_state))
        self.assertTrue(is_valid('O123456789', dl_state))
        self.assertFalse(is_valid('1234567890123', dl_state))

    def test_oregon(self):
        dl_state = 'OR'
        self.assertTrue(is_valid('12345678', dl_state))
        self.assertTrue(is_valid('a123456', dl_state))
        self.assertTrue(is_valid('OR12345', dl_state))
        self.assertFalse(is_valid('a12345678', dl_state))
        self.assertFalse(is_valid('1234567890123', dl_state))

    def test_pennsylvania(self):
        dl_state = 'PA'
        self.assertTrue(is_valid('12345678', dl_state))
        self.assertFalse(is_valid('ab123', dl_state))
        self.assertFalse(is_valid('a123456', dl_state))
        self.assertFalse(is_valid('123456789', dl_state))

    def test_rhode_island(self):
        dl_state = 'RI'
        self.assertTrue(is_valid('1234567', dl_state))
        self.assertFalse(is_valid('ab123', dl_state))
        self.assertTrue(is_valid('v123456', dl_state))
        self.assertTrue(is_valid('V123456', dl_state))
        self.assertFalse(is_valid('1234567890123', dl_state))

    def test_south_carolina(self):
        dl_state = 'SC'
        self.assertTrue(is_valid('12345678', dl_state))
        self.assertTrue(is_valid('1234567890', dl_state))
        self.assertFalse(is_valid('ab123', dl_state))
        self.assertFalse(is_valid('a123456', dl_state))
        self.assertFalse(is_valid('1234567890123', dl_state))

    def test_south_dakota(self):
        dl_state = 'SD'
        self.assertTrue(is_valid('12345678', dl_state))
        self.assertFalse(is_valid('ab123', dl_state))
        self.assertFalse(is_valid('a123456', dl_state))
        self.assertFalse(is_valid('1234567890123', dl_state))

    def test_tennessee(self):
        dl_state = 'TN'
        self.assertTrue(is_valid('1234567', dl_state))
        self.assertTrue(is_valid('123456789', dl_state))
        self.assertFalse(is_valid('ab123', dl_state))
        self.assertFalse(is_valid('a123456', dl_state))
        self.assertFalse(is_valid('1234567890123', dl_state))

    def test_texas(self):
        dl_state = 'TX'
        self.assertTrue(is_valid('12345678', dl_state))
        self.assertFalse(is_valid('1234567', dl_state))
        self.assertFalse(is_valid('ab123', dl_state))
        self.assertFalse(is_valid('a123456', dl_state))
        self.assertFalse(is_valid('1234567890123', dl_state))

    def test_utah(self):
        dl_state = 'UT'
        self.assertTrue(is_valid('1234', dl_state))
        self.assertTrue(is_valid('123456789', dl_state))
        self.assertFalse(is_valid('ab123', dl_state))
        self.assertFalse(is_valid('a123456', dl_state))
        self.assertFalse(is_valid('1234567890123', dl_state))

    def test_vermont(self):
        dl_state = 'VT'
        self.assertTrue(is_valid('12345678', dl_state))
        self.assertTrue(is_valid('1234567a', dl_state))
        self.assertTrue(is_valid('1234567A', dl_state))
        self.assertFalse(is_valid('ab123', dl_state))
        self.assertFalse(is_valid('a123456', dl_state))
        self.assertFalse(is_valid('1234567890123', dl_state))

    def test_virginia(self):
        dl_state = 'VA'
        self.assertTrue(is_valid('123456789', dl_state))
        self.assertTrue(is_valid('a12345678', dl_state))
        self.assertTrue(is_valid('V12345678', dl_state))
        self.assertFalse(is_valid('ab123', dl_state))
        self.assertFalse(is_valid('a123456', dl_state))
        self.assertFalse(is_valid('1234567890123', dl_state))

    def test_washington(self):
        dl_state = 'WA'
        # Valid licenses Old Format WA
        self.assertTrue(is_valid('abcdefg12345', dl_state))
        self.assertTrue(is_valid('ab*****123ab', dl_state))
        self.assertTrue(is_valid('WASH*AB12345', dl_state))
        self.assertTrue(is_valid('ab*INGT1234b', dl_state))
        self.assertTrue(is_valid('WHO**CL99095', dl_state))
        self.assertTrue(is_valid('DOE**CL99095', dl_state))
        self.assertTrue(is_valid('WOO**JT546KA', dl_state))

        # Valid licenses New Format WA
        self.assertTrue(is_valid('WDL3P715863B', dl_state))
        self.assertTrue(is_valid('WDLABCD456DG', dl_state))
        self.assertTrue(is_valid('wdl123456789', dl_state))
        self.assertTrue(is_valid('wdlabcdefghi', dl_state))

        # Invalid Licenses
        self.assertFalse(is_valid('ab123', dl_state))
        self.assertFalse(is_valid('a123456', dl_state))
        self.assertFalse(is_valid('abcd123456789', dl_state))
        self.assertFalse(is_valid('WDLABCD1234', dl_state))

    def test_west_virginia(self):
        dl_state = 'WV'
        self.assertTrue(is_valid('1234567', dl_state))
        self.assertTrue(is_valid('F478888', dl_state))
        self.assertTrue(is_valid('A12345', dl_state))
        self.assertTrue(is_valid('c123456', dl_state))
        self.assertTrue(is_valid('zb12345', dl_state))
        self.assertTrue(is_valid('aZ123456', dl_state))
        self.assertFalse(is_valid('WV1234567', dl_state))
        self.assertFalse(is_valid('a1234567', dl_state))
        self.assertFalse(is_valid('12345', dl_state))
        self.assertFalse(is_valid('1234567890123', dl_state))

    def test_wisconsin(self):
        dl_state = 'WI'
        self.assertTrue(is_valid('a1234567890123', dl_state))
        self.assertTrue(is_valid('W1234567890123', dl_state))
        self.assertFalse(is_valid('ab123', dl_state))
        self.assertFalse(is_valid('a123456', dl_state))
        self.assertFalse(is_valid('1234567890123', dl_state))

    def test_wyoming(self):
        dl_state = 'WY'
        self.assertTrue(is_valid('123456789', dl_state))
        self.assertFalse(is_valid('1234567890', dl_state))
        self.assertFalse(is_valid('ab123', dl_state))
        self.assertFalse(is_valid('a123456', dl_state))
        self.assertFalse(is_valid('1234567890123', dl_state))
        self.assertFalse(is_valid('icantbelieveijustwroteallthesetests',
                                  dl_state))

    def test_territories(self):
        for dl_state in OTHER_STATES:
            self.assertTrue(is_valid('1234567890', dl_state, True))
            self.assertTrue(is_valid('ab123', dl_state, True))
            self.assertTrue(is_valid('**--  ', dl_state, True))
            self.assertTrue(is_valid('.', dl_state, True))
            self.assertTrue(is_valid('icantbelieveijustwroteallthesetests',
                                     dl_state, True))

    def test_invalid_state(self):
        """
        Invalid states should raise an exception
        """
        dl_num = "doesnt_matter"
        self.assertRaises(Exception, is_valid, dl_num, "fakestate")
        self.assertRaises(Exception, is_valid, dl_num, "ZZ")
        self.assertRaises(Exception, is_valid, dl_num, "california")

    def test_allow_territories(self):
        dl_num = "doesnt_matter"
        self.assertRaises(Exception, is_valid, dl_num, "AS", False)
        self.assertRaises(Exception, is_valid, dl_num, "FM", False)
        self.assertRaises(Exception, is_valid, dl_num, "AA", False)
        self.assertTrue(is_valid(dl_num, "AS", True))
        self.assertTrue(is_valid(dl_num, "FM", True))
        self.assertTrue(is_valid(dl_num, "AA", True))


if __name__ == '__main__':
    unittest.main()


dlnvalidation.py
import re


# In alphabetical order by state, not by abbreviation
STATE_FORMATS = {
    'AL': r'^[0-9]{7}$|^[a-z0-9]{6}$',
    'AK': r'^[0-9]{1,7}$',
    'AZ': r'^[abdy][0-9]{8}$|^[0-9]{9}$',
    'AR': r'^[0-9]{8,9}$',
    'CA': r'^[a-z][0-9]{7}$',
    'CO': r'^[0-9]{9}$|^[a-z][0-9]{1,6}$',
    'CT': r'^[0-9]{9}$',
    'DE': r'^[0-9]{1,7}$',
    'DC': r'^[0-9]{7}$|^[0-9]{9}$',
    'FL': r'^[a-z][0-9]{1,12}$',
    'GA': r'^[0-9]{7,9}$',
    'HI': r'^[0-9]{9}$|^H[0-9]{8}$',
    'ID': r'^[0-9]{9}$|^[a-z]{2}[0-9]{6}[a-z]$',
    'IL': r'^[a-z0-9]{12}$',
    'IN': r'^[a-z0-9]{10,11}$',
    'IA': r'^[0-9]{9}$|^[0-9]{3}[a-z]{2}[0-9]{4}$',
    'KS': r'^k[0-9]{8}$|^[0-9]{9}$',
    'KY': r'^[a-z][0-9]{8}$',
    'LA': r'^[0-9]{9}$',
    'ME': r'^[0-9]{7}$',
    'MD': r'^[a-z][0-9]{12}$',
    'MA': r'^s[0-9]{8}$|^[0-9]{9}$',
    'MI': r'^[a-z][0-9]{12}$',
    'MN': r'^[a-z][0-9]{12}$',
    'MS': r'^[0-9]{9}$',
    'MO': r'^[a-z][0-9]{5,9}$|^[0-9]{9}$',
    'MT': r'^[0-9]{13}$|^[a-z]{9}$',
    'NE': r'^[abceghv][0-9]{3,8}$',
    'NV': r'^[0-9]{10}$|^[0-9]{12}$|^x[0-9]{8}$',
    'NH': r'^[0-9]{2}[a-z]{3}[0-9]{5}$',
    'NJ': r'^[a-z][0-9]{14}$',
    'NM': r'^[0-9]{9}$',
    'NY': r'^[a-z][0-9]{18}$|^[0-9]{9}$',
    'NC': r'^[0-9]{1,12}$',
    'ND': r'^[a-z]{3}[0-9]{6}$|^[0-9]{9}$',
    'OH': r'^[a-z]{2}[0-9]{6}$',
    'OK': r'^[a-z][0-9]{9}$|^[0-9]{9}$',
    'OR': r'^[0-9]{1,9}$|^[a-z0-9]{7}$',
    'PA': r'^[0-9]{8}$',
    'RI': r'^[0-9]{7}$|^v[0-9]{6}$',
    'SC': r'^[0-9]{1,10}$',
    'SD': r'^[0-9]{6,9}$',
    'TN': r'^[0-9]{7,9}$',
    'TX': r'^[0-9]{8}$',
    'UT': r'^[0-9]{4,10}$',
    'VT': r'^[0-9]{8}$|^[0-9]{7}[a-z]$',
    'VA': r'^[a-z][0-9]{8}$|^[0-9]{9}$',
    'WA': r'^[a-z*]{7}[0-9]{3}[0-9a-z]{2}$',
    'WV': r'^[0-9]{7}$|^[a-z]{1,2}[0-9]{5,6}$',
    'WI': r'^[a-z][0-9]{13}$',
    'WY': r'^[0-9]{9}$'
}


OTHER_STATES = [
    # Territorries
    'AS', 'GU', 'MP', 'PR', 'VI', 'UM',

    # Freely Associated States
    'FM', 'MH', 'PW',

    # Armed Forces
    'AA', 'AE', 'AP'
]


def is_valid(dl_number, dl_state, allow_territories=False):
    """
    dl_number is a drivers license number string.
    dl_state is the state abbreviation of the state of the drivers license.
    if allow_territories is set, all drivers licenses for territorries,
    freely associated states and armed forces will be processed as valid.
    If not set, they will be considered an invalid state and an exception
    will be thrown.
    Returns True if given a valid drivers license number.

    All drivers license numbers must be alphanumeric and conform to
    specific state formats.
    For more info, see http://adr-inc.com/PDFs/State_DLFormats.pdf.
    """

    if not dl_number:
        return False

    if dl_state in STATE_FORMATS:
        regex = STATE_FORMATS[dl_state]
        is_valid = bool(re.match(regex, dl_number, re.IGNORECASE))
        return is_valid
    elif allow_territories and dl_state in OTHER_STATES:
        return True
    else:
        raise Exception("Invalid state for drivers license number validation!")


dl_number_validation.py
import re


# In alphabetical order by state, not by abbreviation
STATE_FORMATS = {
    'AL': r'[0-9]{1,7}$',
    'AK': r'[0-9]{1,7}$',
    'AZ': r'[a-z][0-9]{1,8}$|[a-z]{2}[0-9]{2,5}$|[0-9]{9}$',
    'AR': r'[0-9]{4,9}$',
    'CA': r'[a-z][0-9]{7}$',
    'CO': r'[0-9]{9}$|[a-z][0-9]{3,6}$|[a-z]{2}[0-9]{2,5}$',
    'CT': r'[0-9]{9}$',
    'DE': r'[0-9]{1,7}$',
    'DC': r'[0-9]{7}$|[0-9]{9}$',
    'FL': r'[a-z][0-9]{12}$',
    'GA': r'[0-9]{7,9}$',
    'HI': r'[a-z][0-9]{8}$|[0-9]{9}$',
    'ID': r'[a-z]{2}[0-9]{6}[a-z]$|[0-9]{9}$',
    'IL': r'[a-z][0-9]{11,12}$',
    'IN': r'[a-z][0-9]{9}$|[0-9]{9,10}$',
    'IA': r'[0-9]{9}$|[0-9]{3}[a-z]{2}[0-9]{4}$',
    'KS': r'[a-z][0-9][a-z][0-9][a-z]|[a-z][0-9]{8}$|[0-9]{9}$',
    'KY': r'[a-z][0-9]{8,9}$|[0-9]{9}$',
    'LA': r'[0-9]{1,9}$',
    'ME': r'[0-9]{7,8}$|[0-9]{7}[a-z]$',
    'MD': r'[a-z][0-9]{12}$',
    'MA': r'[a-z][0-9]{8}$|[0-9]{9}$',
    'MI': r'[a-z]([0-9]{10}|[0-9]{12})$',
    'MN': r'[a-z][0-9]{12}$',
    'MS': r'[0-9]{9}$',
    'MO': r'[a-z][0-9]{5,9}$|[a-z][0-9]{6}r$|[0-9]{8}[a-z]{2}$|[0-9]{9}[a-z]$|[0-9]{9}$',
    'MT': r'[a-z][0-9]{8}$|[0-9]{13,14}$|[0-9]{9}$',
    'NE': r'[0-9]{1,7}$',
    'NV': r'[0-9]{9,10}$|[0-9]{12}$|x[0-9]{8}$',
    'NH': r'[0-9]{2}[a-z]{3}[0-9]{5}$',
    'NJ': r'[a-z][0-9]{14}$',
    'NM': r'[0-9]{8,9}$',
    'NY': r'[a-z]([0-9]{7}|[0-9]{18})$|[0-9]{8,9}$|[0-9]{16}$|[a-z]{8}$',
    'NC': r'[0-9]{1,12}$',
    'ND': r'[a-z]{3}[0-9]{6}$|[0-9]{9}$',
    'OH': r'[a-z][0-9]{4,8}$|[a-z]{2}[0-9]{3,7}$|[0-9]{8}$',
    'OK': r'[a-z][0-9]{9}$|[0-9]{9}$',
    'OR': r'[0-9]{1,9}$|[a-z][0-9]{6}$|[a-z]{2}[0-9]{5}$',
    'PA': r'[0-9]{8}$',
    'RI': r'[0-9]{7}$|[a-z][0-9]{6}$',
    'SC': r'[0-9]{5,11}$',
    'SD': r'[0-9]{6,10}$|[0-9]{12}$',
    'TN': r'[0-9]{7,9}$',
    'TX': r'[0-9]{7,8}$',
    'UT': r'[0-9]{4,10}$',
    'VT': r'[0-9]{8}$|[0-9]{7}a$',
    'VA': r'[a-z][0-9]{9,11}$|[0-9]{9}$',
    'WA': r'[a-z][a-z0-9*]{11}$',
    'WV': r'[0-9]{7}$|[a-z]{1,2}[0-9]{5,6}$',
    'WI': r'[a-z][0-9]{13}$',
    'WY': r'[0-9]{9,10}$'
}


OTHER_STATES = [
    # Territorries
    'AS', 'GU', 'MP', 'PR', 'VI', 'UM',

    # Freely Associated States
    'FM', 'MH', 'PW',

    # Armed Forces
    'AA', 'AE', 'AP'
]


def is_valid(dl_number, dl_state, allow_territories=False):
    """
    dl_number is a drivers license number string.
    dl_state is the state abbreviation of the state of the drivers license.
    if allow_territories is set, all drivers licenses for territorries,
    freely associated states and armed forces will be processed as valid.
    If not set, they will be considered an invalid state and an exception
    will be thrown.
    Returns True if given a valid drivers license number.

    All drivers license numbers must be alphanumeric and conform to
    specific state formats.
    For more info, see http://adr-inc.com/PDFs/State_DLFormats.pdf.
    """

    if not dl_number:
        return False

    if dl_state in STATE_FORMATS:
        regex = STATE_FORMATS[dl_state]
        is_valid = bool(re.match(regex, dl_number, re.IGNORECASE))
        return is_valid
    elif allow_territories and dl_state in OTHER_STATES:
        return True
    else:
        raise Exception("Invalid state for drivers license number validation!")
